File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/deps/conflict.py
from __future__ import annotations

from .range import intersects, parse_range


def check_conflicts(registry) -> list:
    grouped: dict = {}
    for caller, deps in registry.edges.items():
        for dep in deps:
            if dep.origin != "in_repo" or not dep.version_range:
                continue
            grouped.setdefault(dep.name, []).append(
                {
                    "caller": caller,
                    "range_str": dep.version_range,
                    "range": parse_range(dep.version_range),
                }
            )
    issues: list = []
    for dep_name in sorted(grouped.keys()):
        callers = grouped[dep_name]
        for i in range(len(callers)):
            for j in range(i + 1, len(callers)):
                if not intersects(callers[i]["range"], callers[j]["range"]):
                    issues.append(
                        {
                            "code": "E020_CONFLICTING_RANGES",
                            "severity": "error",
                            "field": "dependencies",
                            "message": (
                                f"incompatible ranges for '{dep_name}': "
                                f"{callers[i]['caller']} requires {callers[i]['range_str']}, "
                                f"{callers[j]['caller']} requires {callers[j]['range_str']}"
                            ),
                        }
                    )
    return issues


def check_unpinned(registry) -> list:
    issues: list = []
    for caller, deps in registry.edges.items():
        for dep in deps:
            if dep.origin == "in_repo" and not dep.version_range:
                issues.append(
                    {
                        "code": "W020_UNPINNED_DEPENDENCY",
                        "severity": "warn",
                        "field": "dependencies",
                        "message": (
                            f"{caller}: unpinned dependency '{dep.name}'; "
                            f"pin against tag '{dep.name}/v<X.Y.Z>'"
                        ),
                    }
                )
    return issues

File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/deps/conflict.mjs
import { parseRange, intersects } from "./range.mjs";

/**
 * For each in-repo dependency name, collect every (caller, range) pair.
 * If any two ranges fail `intersects`, emit E020_CONFLICTING_RANGES.
 * Issues are returned in deterministic order (sorted by message).
 */
export function checkConflicts(registry) {
  const grouped = new Map();
  // Iterate edges in sorted-by-caller order; registry already returns a sorted Map.
  for (const [caller, deps] of registry.edges) {
    for (const dep of deps) {
      if (dep.origin !== "in_repo" || !dep.version_range) continue;
      const arr = grouped.get(dep.name) ?? [];
      arr.push({
        caller,
        rangeStr: dep.version_range,
        range: parseRange(dep.version_range),
      });
      grouped.set(dep.name, arr);
    }
  }
  const issues = [];
  // Sort group iteration for determinism
  const groups = [...grouped.entries()].sort(([a], [b]) =>
    a < b ? -1 : a > b ? 1 : 0,
  );
  for (const [depName, callers] of groups) {
    for (let i = 0; i < callers.length; i++) {
      for (let j = i + 1; j < callers.length; j++) {
        if (!intersects(callers[i].range, callers[j].range)) {
          issues.push({
            code: "E020_CONFLICTING_RANGES",
            severity: "error",
            field: "dependencies",
            message: `incompatible ranges for '${depName}': ${callers[i].caller} requires ${callers[i].rangeStr}, ${callers[j].caller} requires ${callers[j].rangeStr}`,
          });
        }
      }
    }
  }
  return issues;
}

/**
 * Emit W020_UNPINNED_DEPENDENCY for each in-repo dep that has no version range.
 */
export function checkUnpinned(registry) {
  const issues = [];
  for (const [caller, deps] of registry.edges) {
    for (const dep of deps) {
      if (dep.origin === "in_repo" && !dep.version_range) {
        issues.push({
          code: "W020_UNPINNED_DEPENDENCY",
          severity: "warn",
          field: "dependencies",
          message: `${caller}: unpinned dependency '${dep.name}'; pin against tag '${dep.name}/v<X.Y.Z>'`,
        });
      }
    }
  }
  return issues;
}
